### PR TITLE
Introduced Codec

### PR DESCRIPTION
**Pull request by mairbek**
Codecs handle encoding and decoding of gRPC messages.

Comes with two types of codecs, `GRPC.Codec.Proto`(default) and `GRPC.Codec.Erlpack`.

Currently, codecs are specified as an option on `Server` and `Stub`. 

Looking at the gRPC protocol and implementations in other languages, there should be a way to also specify a codec per request. That will require us registering multiple codecs, and being able to override a codec on the individual call, for example

```elixir
Helloworld.Greeter.Stub.reply_hello(name, codec: GRPC.Codec.Erlpack)
```

I'm still iterating on it but it would be great to get some early feedback. We've found this refactoring useful because we could start using the library without rewriting our code to use protos and we can stick with erlang serialization for a little while.

### CODE REVIEW

**Comment by mairbek** [comment]
https://github.com/elixir-grpc/grpc/issues/37 should be possible to add json and flatbuffers eventually

File: lib/grpc/erlpack_codec.ex [hunk 1]
```diff
@@ -0,0 +1,15 @@
+defmodule GRPC.Codec.Erlpack do
```
Comment (tony612) [inline lib/grpc/erlpack_codec.ex]: Could you put this file in grpc/codec/erlpck.ex, as well protobuf

**Comment by tony612** [comment]
I saw you added codec to stream. Could we have a connection level codec also?

In addition, though I saw grpc-go supports stream level codec, but I'm thinking if we really need a stream level codec. If people don't use this often, I suggest we can only add a connection level.

**Comment by mairbek** [comment]
go-grpc has a `https://godoc.org/google.golang.org/grpc#WithDefaultCallOptions` and allows to override a codec during the call.

I'm going to model the code accordingly.

**Comment by mairbek** [comment]
I've tweaked the code a little bit

* Rather than having one codec, a server can support a list of codecs
* The server picks reads the request content type and picks the codec accordingly 
* The client can specify a `codec` option when sending a request
* The client can specify a default channel coding

File: lib/grpc/server.ex [hunk 2]
```diff
@@ -40,14 +40,20 @@ defmodule GRPC.Server do
...
+      service_mod = opts[:service]
+      codec = opts[:codec]
       service_name = service_mod.__meta__(:name)
 
       Enum.each(service_mod.__rpc_calls__, fn {name, _, _} = rpc ->
         func_name = name |> to_string |> Macro.underscore() |> String.to_atom()
         path = "/#{service_name}/#{name}"
         grpc_type = GRPC.Service.grpc_type(rpc)
 
+        def get_codec() do
```
[the remark this answers is not in the document]
Comment (mairbek) [reply lib/grpc/server.ex]: I don't really know how to make this better in a way that the code compiles..

[the remark this answers is not in the document]
Comment (tony612) [reply lib/grpc/server.ex]: Just put code between `defmacro` and `quote`

**Comment by tony612** [comment]
@mairbek Is this finished?

**Comment by mairbek** [comment]
@tony612 should be good now, please take a look

File: lib/grpc/server.ex (hunk 2 above)
[the remark this answers is not in the document]
Comment (mairbek) [reply lib/grpc/server.ex]: Done

**Comment by mairbek** [comment]
@tony612 I've updated the codec code in `lib/grpc/server.ex`

**Comment by tony612** [comment]
I merged and and made some changes, like changing `@callback content_subtype` to `@callback name`.

https://github.com/elixir-grpc/grpc/commit/2334e933dc585660a7c3096886c6f6e2c752dbda#diff-a29ed486abbca54464dcc8839c13b754R7

**Pull request merged**